plugins/poetry_env/poetry_env.plugin.zsh

```diff
@@ -0,0 +1,24 @@
+# Automatic poetry shell activation/deactivation
+_togglePoetryShell() {
+  # deactivate shell if pyproject.toml doesn't exist and not in a subdir
+  if [[ ! -f "$PWD/pyproject.toml" ]] ; then
+    if [[ "$POETRY_ACTIVE" == 1 ]]; then
+      if [[ "$PWD" != "$poetry_dir"* ]]; then
+        exit
+      fi
+    fi
+  fi
+
+  # activate the shell if pyproject.toml exists
+  if [[ "$POETRY_ACTIVE" != 1 ]]; then
+    if [[ -f "$PWD/pyproject.toml" ]]; then
+      if grep -q 'tool.poetry' "$PWD/pyproject.toml"; then
+        export poetry_dir="$PWD"
+        source "$( poetry env list --full-path | grep Activated | cut -d' ' -f1 )/bin/activate"
```

Comment:
@hamedsh The changes don't work as expected. 
1. To activate poetry shell you have to run the command `poetry shell`. Sourcing virtual env activate file doesn't activate a new shell but sources the environment in current shell. Sourcing also doesn't set the environment variable POETRY_ACTIVE, it is only set in `poetry shell`.
2. Even if you activate the shell using `poetry shell`, deactivation doesn't work as expected since the chpwd hook exits the current shell (which is poetry shell) and cwd in the outer shell is not changed (it is still the project directory).

I am not a zsh expert so not sure if there is something you can do in the hook to resolve point no. 2. 

I suggest another alternative to activate/deactivate poetry environment as -

```zsh
# Automatic poetry environment activation/deactivation
_togglePoetryEnv() {
  # deactivate shell if pyproject.toml doesn't exist and not in a subdir
  if [[ ! -f "$PWD/pyproject.toml" ]] ; then
    if [[ "$poetry_active" == 1 ]]; then
      if [[ "$PWD" != "$poetry_dir"* ]]; then
        export poetry_active=0
        deactivate
        return
      fi
    fi
  fi

  # activate the environment if pyproject.toml exists
  if [[ "$poetry_active" != 1 ]]; then
    if [[ -f "$PWD/pyproject.toml" ]]; then
      if grep -q 'tool.poetry' "$PWD/pyproject.toml"; then
        export poetry_active=1
        export poetry_dir="$PWD"
        source "$(poetry env info --path)/bin/activate"
      fi
    fi
  fi
}
autoload -U add-zsh-hook
add-zsh-hook chpwd _togglePoetryEnv
_togglePoetryEnv
```